File: leetcode/medium/ReorderLinkedList.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Reorders a singly linked list in-place.
        
        Given a list L0 -> L1 -> ... -> Ln-1 -> Ln, it reorders it to
        L0 -> Ln -> L1 -> Ln-1 -> L2 -> Ln-2 -> ...

        The solution follows three steps:
        1. Find the middle of the linked list to split it into two halves.
        2. Reverse the second half of the list.
        3. Merge the first half and the reversed second half.
        
        This is done in O(N) time and O(1) space.

        Args:
            head: The head of the linked list. The list is modified in-place.
        """
        if not head or not head.next:
            return

        # Step 1: Find the middle of the linked list.
        # 'slow' will point to the end of the first half.
        # 'fast' is used to find the middle.
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Split the list into two halves.
        # 'second_head' is the head of the second half.
        second_head = slow.next
        slow.next = None

        # Step 2: Reverse the second half of the list.
        prev = None
        curr = second_head
        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        # 'prev' is now the new head of the reversed second half.
        second_head = prev

        # Step 3: Merge the first and the reversed second halves.
        first_head = head
        while second_head:
            # Save the next nodes of both lists before modifying pointers.
            temp1 = first_head.next
            temp2 = second_head.next

            # Interleave the nodes.
            first_head.next = second_head
            second_head.next = temp1

            # Move to the next pair of nodes to be merged.
            first_head = temp1
            second_head = temp2

File: leetcode/medium/ReorderLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    /**
     * Reorders the singly-linked list in-place to the format L0 -> Ln -> L1 -> Ln-1 -> L2 -> Ln-2 -> ...
     * The process involves three main steps:
     * 1. Find the middle of the linked list using the Tortoise and Hare (slow/fast pointer) approach.
     * 2. Reverse the second half of the list.
     * 3. Merge the first half and the reversed second half alternately.
     *
     * Time Complexity: O(n) - Each of the three steps traverses the list (or half of it) once.
     * Space Complexity: O(1) - All operations are done in-place with a constant number of extra pointers.
     *
     * @param head The head of the singly-linked list to be reordered.
     */
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return;
        }

        // --- Step 1: Find the middle of the linked list ---
        // 'slow' will stop at the node just before the start of the second half
        // For a list 1->2->3->4, slow stops at 2. For 1->2->3->4->5, slow stops at 3.
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        // 'secondHead' is the start of the second half.
        ListNode secondHalfHead = slow.next;
        // Break the list into two halves.
        slow.next = null;
        
        

        // --- Step 2: Reverse the second half of the list ---
        ListNode current = secondHalfHead;
        ListNode previous = null;
        
        while (current != null) {
            ListNode nextNode = current.next;
            current.next = previous; // Reverse the link
            previous = current;      // Move previous one step forward
            current = nextNode;      // Move current one step forward
        }
        
        // 'secondHalfHead' now points to the head of the reversed second half (Ln)
        secondHalfHead = previous;

        

        // --- Step 3: Merge the two halves alternately ---
        ListNode firstHalfCurrent = head; // L0, L1, L2, ...
        ListNode secondHalfCurrent = secondHalfHead; // Ln, Ln-1, Ln-2, ...
        
        while (secondHalfCurrent != null) {
            // Save the next nodes before overwriting links
            ListNode firstHalfNext = firstHalfCurrent.next;
            ListNode secondHalfNext = secondHalfCurrent.next;

            // Reorder: first -> second
            firstHalfCurrent.next = secondHalfCurrent;
            // Reorder: second -> first's next
            secondHalfCurrent.next = firstHalfNext;

            // Move pointers to the saved next nodes
            firstHalfCurrent = firstHalfNext;
            secondHalfCurrent = secondHalfNext;
        }
        // Note: The loop terminates when secondHalfCurrent is null, meaning all nodes
        // from the second half have been merged. The first half pointer (firstHalfCurrent)
        // will naturally be null or point to the node whose 'next' is null (if the list had an odd length).
        // Since 'slow.next' was set to null in Step 1, the new tail of the list is correctly set.
    }
}